src/duplex.rs

```diff
@@ -50,71 +73,73 @@ impl AudioInputCallback for InputProxy {
     fn on_input_data(&mut self, context: AudioCallbackContext, input: AudioInput<f32>) {
         log::trace!(num_samples = input.buffer.num_samples(), num_channels = input.buffer.num_channels();
             "on_input_data");
-        let input_slots = input.buffer.num_samples() * input.buffer.num_channels();
-        if self.buffer.slots() < input_slots {
-            log::error!(buffer_slots=self.buffer.slots(), input_slots; "Input proxy buffer underrun");
-            return;
-        }
-        let mut scratch = [0f32; 32];
-        let rate = self.output_sample_rate.load(Ordering::SeqCst) as f64
-            / context.stream_config.samplerate;
-        let out_len = (input.buffer.num_samples() as f64 * rate) as usize;
-        let mut scratch =
-            ArrayViewMut1::from(&mut scratch[..context.stream_config.channels.count()]);
-        let rate_recip = rate.recip();
-        for i in 0..out_len {
-            let in_ix = i as f64 / rate_recip;
-            let i = in_ix.floor() as usize;
-            let j = i + 1;
-            if j == out_len {
-                scratch.assign(&input.buffer.get_frame(i));
-            } else {
-                lerp(
-                    in_ix.fract() as _,
-                    input.buffer.get_frame(i),
-                    input.buffer.get_frame(j),
-                    scratch.view_mut(),
-                );
-            }
-            for sample in scratch.iter().copied() {
-                let _ = self.buffer.push(sample);
+        if let Ok(output_samplerate) = self.receive_output_samplerate.pop() {
+            let Some(num_channels) = NonZeroUsize::new(context.stream_config.channels.count())
+            else {
+                log::error!("Input proxy: no input channels given");
+                return;
+            };
+            let input_samplerate = context.stream_config.samplerate as _;
+            log::debug!(
+                "Creating resampling channel ({} Hz) -> ({} Hz) ({} channels)",
+                input_samplerate,
+                output_samplerate,
+                num_channels.get()
+            );
+            let (tx, rx) = fixed_resample::resampling_channel(
+                num_channels,
+                input_samplerate,
+                output_samplerate,
+                ResamplingChannelConfig {
+                    latency_seconds: 0.01,
+                    quality: fixed_resample::ResampleQuality::Low,
```

Comment:
Later I see that this is configurable. Does it not need to be configurable here?

Reply:
This is a question I've been asking myself -- should we expose the configuration or should we make this seamless to the user? There's an argument for providing good defaults but exposing the setting, but it might also be confusing to have those options available.

I myself don't know if I can set the `latency_seconds` to zero, theoretically I should be able to and `fixed_resample` will try to get the lowest latency it can, but also it might just break, and exposing that to the user feels like exposing footguns.

I think I'm leaning towards providing the configuration with good defaults, but not for latency which if it works I'll keep at 0 because I don't think it makes sense to have anything else (I'll probably ask on Discord about it though).